src/GraphQL/Resolver.hs

```diff
@@ -379,30 +382,33 @@ instance forall m a b dispatchType.
          , KnownSymbol (FieldName dispatchType)
          , Monad m
          ) => RunFields m (a :> b) where
-  runFields handler field =
-    resolveField @dispatchType @m handler nextHandler field
+  runFields allTypes handler field =
+    resolveField @dispatchType @m allTypes handler nextHandler field
     where
       nextHandler = pure (Result [FieldNotFoundError (getName field)] Nothing)
 
 instance forall typeName interfaces fields m.
          ( RunFields m (RunFieldsType m fields)
+         , API.HasObjectDefinition (API.Object typeName interfaces fields)
          , Monad m
          ) => HasResolver m (API.Object typeName interfaces fields) where
   type Handler m (API.Object typeName interfaces fields) = m (RunFieldsHandler m (RunFieldsType m fields))
 
-  resolve mHandler selectionSet = do
-    -- First we run the actual handler function itself in IO.
-    handler <- mHandler
-    let fields = getFields selectionSet
-    -- We're evaluating an Object so we're collecting ObjectFields from
-    -- runFields and build a GValue.Map with them.
-    r <- forM fields (runFields @m @(RunFieldsType m fields) handler)
-    -- let (errs, fields) = foldr' (\(Result ea fa) (eb, fbs) -> (eb <> ea, fa:fbs)) ([], []) r
-    let (Result errs obj)  = GValue.makeObject . catMaybes <$> sequenceA r
-    case obj of
-      Nothing -> pure (Result [DuplicateFields r] GValue.ValueNull)
-      Just object -> pure (Result errs (GValue.ValueObject object))
+  resolve _ _ Nothing = throwE MissingSelectionSet
+  resolve allTypes mHandler (Just selectionSet) =
+    case getSelectionSet of
+      Left err -> throwE err
+      Right ss -> do
+        handler <- mHandler
+        r <- traverse (runFields @m @(RunFieldsType m fields) allTypes handler) ss
+        let (Result errs obj)  = GValue.objectFromOrderedMap . OrderedMap.catMaybes <$> sequenceA r
+        pure (Result errs (GValue.ValueObject obj))
 
+    where
+      getSelectionSet = do
+        defn <- first SchemaError $ API.getDefinition @(API.Object typeName interfaces fields)
+        (SelectionSet ss') <- first ValidationError $ getSelectionSetForType (flip Map.lookup allTypes) defn selectionSet
```

Comment:
Something like...

```haskell
-- Fields of a selection set may be behind "type conditions", due to inline fragments 
-- or the use of fragment spreads. These type conditions are represented in the 
-- schema by the name of a type (e.g. "Dog"). To determine which type conditions 
-- (and thus which fields) are relevant for this selection set, we need to look up the 
-- actual types they refer to, as interfaces (say) match objects differently than 
-- unions.
-- 
-- See <https://facebook.github.io/graphql/#sec-Field-Collection> for more details.
```

Can you please provide similar levels of commentary on literally everything in Resolver.hs?